sensorium/training/trainers.py

```diff
@@ -168,6 +252,40 @@ def full_objective(model, dataloader, data_key, *args, **kwargs):
         lr_decay_steps=lr_decay_steps,
     ):
 
+        if include_kldivergence and epoch == dec_starting_epoch:
+            # TODO: include hidden dimension
+            kmeans = KMeans(
+                n_clusters=cluster_number, n_init=kmeans_init, random_state=seed
+            )
+            model.train()
+            feature_list = []
+            features_subset = []
+            random_indices = {}
+            # form initial cluster centres
+
+            for k, readout in model.readout.items():
+                features = readout.features.cpu().detach().squeeze().T.numpy()
+                feature_list.append(np.array(features))
+                random_index = np.random.choice(
+                    features.shape[0], size=int(features.shape[0] / 5), replace=False
+                )
+                random_indices[k] = random_index
+                features_subset.append(features[random_indices[k], :])
+            features = np.vstack(feature_list)
+            features_subset = np.vstack(features_subset)
+            predicted = kmeans.fit_predict(features)
```

Comment:
I think this is initialization, so there should be no gradients through this stage
To make it more explicit you can do sth like
```suggestion
            with torch.no_grad():
                for k, readout in model.readout.items():
                    features = readout.features.cpu().detach().squeeze().T.numpy()
                    feature_list.append(np.array(features))
                    random_index = np.random.choice(
                        features.shape[0], size=int(features.shape[0] / 5), replace=False
                    )
                    random_indices[k] = random_index
                    features_subset.append(features[random_indices[k], :])
                features = np.vstack(feature_list)
                features_subset = np.vstack(features_subset)
                predicted = kmeans.fit_predict(features)
```

Reply:
True. Thank you